### DISCUSSION TITLE
TabBar: Ability to make tabs expand to fill blank space

### DISCUSSION THREAD

**Discussion by devvoid** [godotengine/godot-proposals#8397] [GUI]
It would be extremely useful if there were some way of making tabs expand to fill all empty space in a TabBar, in a way similar to a control's "expand" size flag.

Here's an example of what I mean: On the left is a standard TabContainer, while on the right is a hacked-together mockup I made, which is an HBoxContainer full of buttons with their horizontal size flags set to "expand".
![image](https://github.com/godotengine/godot-proposals/assets/33298798/fe2532f6-d388-4377-9e02-2a3cdb930ecd)

This looks much better visually, especially in situations where there's a fixed number of tabs and the amount will never change, but obviously to implement this in an actual game would take a lot of re-implementing TabContainer's functionality. It would be incredibly useful if there were just a property you could set to make this happen - maybe another entry on the `AlignmentMode` enum?

**Comment by feois** [comment]
I've no idea why isn't this a thing. Anyway currently to implement it by code?

**Reply by bhottovy** [reply]
I don't think so, not without manually calling "set_tab_title" and padding the title with whitespace to get the correct size.

**Comment by bhottovy** [comment]
I second that this would be a very useful feature, it seems that everything related to tab sizing is done internally and there is no way to really override it. At the very least I wish there was an option to set a minimum width for the tabs.

**Comment by TheZomb** [comment]
+1

**Reply by Calinou** [reply]
@TheZomb Please don't bump issues without contributing significant new information. Use the :+1: reaction button on the first post instead.

**Comment by davthedev** [comment]
Probably doable, will have a look to try to implement it

**Reply by TheZomb** [reply]
@davthedev I've currently already started implementing it, but would be glad if we could work together on it. That would make both our lives easier. Let's get into contact via mail

**Reply by davthedev** [reply]
To kill two birds with one stone, we can use this opportunity to implement an actual "tab alignment" setting.

![image](https://github.com/godotengine/godot-proposals/assets/26961646/f9bc266e-5f97-465b-b08f-808fd38c83c2)

**Reply by TheZomb** [reply]
yeah that sounds great. and it should also work fine with the function of having tabs larger than the tab bar width hidden and cyclable.  All in all, the only thing that needs to be implemented actually is being able to set tab min width I think. Everything else should be doable with the currently implemented features

**Reply by taelspinner** [reply]
@TheZomb @davthedev Any reason this didn't get implemented?

**Reply by taelspinner** [reply]
Ah, I see it was implemented partially: https://github.com/godotengine/godot/pull/103214

Unfortunately, this specific feature (tab expansion to fill) did not make the cut.

**Reply by aaronjyoder** [reply]
Will take a look at implementing sometime in October, unless @TheZomb and @davthedev return with their own implementation.

**Reply by aaronjyoder** [reply]
Additionally, I'd like some clarity on implementation requirements. Is a user-facing minimum tab width option a requirement of this proposal? Adding an option to expand tab width to fill the container would not necessarily require presenting a user-facing option to set a minimum tab width.

I'd also be somewhat concerned about exactly how this would look in the Godot editor to have both options, as minimum tab width conflicts directly with expand tab width to fill container, and it's not intuitive that one of the alignment options would override a minimum tab width setting. If we have a minimum tab width setting, would we also want a maximum tab width setting?

I'm leaning toward an "expand tab width" option NOT being an alignment option, as it's really unrelated to alignment, as the uniform sizing shows above. It would simply mean alignment would have no effect. Justify does make sense as an alignment option, however.

**Reply by taelspinner** [reply]
I would say having a horizontal fill option would be the best implementation if you wanted to support the behaviors mocked up above, with four options: Fit (current behavior), Uniform (tabs expand to have equal size based on the widest tab), Fill (tabs expand to evenly fill the width), and Justify (tabs expand to fill the width proportionally based on their original width).

**Reply by aaronjyoder** [reply]
Hey everyone, just a quick update: the ongoing government shutdown in the US has impacted my ability to spend time working on this, which is why nothing has happened in the past month.

I'm working on it a bit tonight, so hoping to have some kind of substantial update soon-ish.

**Reply by aaronjyoder** [reply]
Hey everyone! I've got a pull request open [here](https://github.com/godotengine/godot/pull/113385). It very likely won't make it for the 4.6 feature freeze cut-off date (December 3) but it's out there now!

~~It'll likely need tests to be written for it, since I was only given a pass on tests for [#103214](https://github.com/godotengine/godot/pull/103214) due to it being my first PR. I don't have time to write the tests right now, so if someone else wanted to pitch in and get tests written for both [#103214](https://github.com/godotengine/godot/pull/103214) and [#113385](https://github.com/godotengine/godot/pull/113385) in the second PR, I'd much appreciate it! Otherwise, I'll get to it when I can.~~ Tests written.

**Comment by TiptoeAndDoomsday** [comment]
If you put this code in a Tab Container it should add space to both sides of your Tabs:

[https://gist.github.com/TiptoeAndDoomsday/10131d8208fd7847fc5aa159a8544569](https://gist.github.com/TiptoeAndDoomsday/10131d8208fd7847fc5aa159a8544569)

**Comment by aaronjyoder** [comment]
Implemented in this PR: https://github.com/godotengine/godot/pull/113385
